### PR TITLE
Support for http based grpc-web for mobile

### PR DESCRIPTION
**Pull request by otakuto**
https://github.com/grpc/grpc-dart/pull/287
https://github.com/grpc/grpc-dart/pull/351
This will be a re-attempting of the above.

Thanks for your review.

### CODE REVIEW

**Comment by devoncarew** [comment]
@sigurdm @mraleph - who would be best to review this?

**Comment by mraleph** [comment]
@otakuto thanks for taking your time to contribute. Could you try to fix EasyCLA check? We can't look at the code before it is fixed.

**Comment by mraleph** [comment]
@otakuto Could you outline a specific use case which makes you interested in using gRPC-Web from a mobile application? Any specific reason that prevents you from using the native gRPC protocol instead?

/cc @wenbozhu

**Comment by otakuto** [comment]
> @otakuto thanks for taking your time to contribute. Could you try to fix EasyCLA check? We can't look at the code before it is fixed.

I fixed it.

**Comment by otakuto** [comment]
> @otakuto Could you outline a specific use case which makes you interested in using gRPC-Web from a mobile application? Any specific reason that prevents you from using the native gRPC protocol instead?
> 
> /cc @wenbozhu

In this example, there is a reason why it is difficult for a microcontroller to support HTTP2 when communicating with a microcontroller such as ESP32 and mobile, for example.
Another advantage is that it is possible to implement a pure dart client for mobile and web respectively.

**Comment by mraleph** [comment]
@otakuto 

> In this example, there is a reason why it is difficult for a microcontroller to support HTTP2 when communicating with a microcontroller such as ESP32 and mobile.

Makes sense. HTTP2 stack might be quite large so it does not necessarily fit on a microcontroller. 

> Another advantage is that it is possible to implement a pure dart client for mobile and web respectively.

Could you elaborate what you mean by this?

One of the problems with merging this (and officially supporting gRPC-Web protocol on native targets) is that it would block us from migrating our gRPC-Web implementation to use official JavaScript implementation underneath, which is something that we have been considering to reduce our maintenance burden.

**Comment by johynpapin** [comment]
Hello !

I want to share another use case for this feature.

As written in #216, another use case is to be able to use the gRPC-Web protocol to bypass the limitations of some cloud platforms, such as DigitalOcean's App Platform service which does not allow to expose a gRPC service.

**Comment by mreimer** [comment]
I'd like to add another vote in support of this PR. AWS lambda doesn't support HTTP2, and while I'm not familiar with other FaaS environments I expect lambda isn't the only one with this constraint. I appreciate the need to control your maintenance surface area but I think it would be a shame to commit to never supporting grpc-web on native targets.

For what it's worth I've locally rebased this PR on 3.1.0 and tested it for my use case, LGTM.

**Comment by johynpapin** [comment]
Hello !

We have tested these changes on our API and everything seems OK on our side too.

Maybe another solution would be to export what is needed to be able to offer this functionality in another package?

File: lib/src/client/transport/http_transport.dart [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  final StreamController<GrpcMessage> _incomingMessages = StreamController();
+  final StreamController<List<int>> _outgoingMessages = StreamController();
+
+  @override
+  Stream<GrpcMessage> get incomingMessages => _incomingMessages.stream;
+
+  @override
+  StreamSink<List<int>> get outgoingMessages => _outgoingMessages.sink;
+
+  HTTPTransportStream(client, request,
```
Comment (mosuem) [inline lib/src/client/transport/http_transport.dart:44]: The arguments should be typed.

File: lib/src/client/transport/http_transport.dart [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+import '../../shared/message.dart';
+import '../../shared/status.dart';
+import '../connection.dart';
+import 'cors.dart' as cors;
+import 'transport.dart';
+import 'web_streams.dart';
+
+const _contentTypeKey = 'Content-Type';
+
+class HTTPTransportStream implements GrpcTransportStream {
```
Comment (mosuem) [inline lib/src/client/transport/http_transport.dart:31]: Rename this and similar to ``Http...` for consistency.

File: lib/src/client/transport/http_transport.dart [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    _onDone(this);
+  }
+
+  @override
+  Future<void> terminate() async {
+    _close();
+  }
+}
+
+class HTTPClientConnection extends ClientConnection {
```
Comment (mosuem) [inline lib/src/client/transport/http_transport.dart:110]: Why `extend` and not `implement`?

File: lib/src/client/transport/http_transport.dart [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    if (callOptions is WebCallOptions &&
+        callOptions.bypassCorsPreflight == true) {
+      requestUri = cors.moveHttpHeadersToQueryParam(metadata, requestUri);
+    }
+
+    final request = http.StreamedRequest('POST', requestUri);
+
+    for (final header in metadata.keys) {
+      request.headers[header] = metadata[header]!;
+    }
```
Comment (mosuem) [inline lib/src/client/transport/http_transport.dart:146]: Should the other headers also be set as in the `XhrClientConnection`?

File: lib/src/client/transport/http_transport.dart [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        }
+        _incomingProcessor.add(Uint8List.fromList(event).buffer);
+      }, onDone: () {
+        _onRequestDone(response);
+        _close();
+      }, onError: (error, stackTrace) {
+        if (_incomingProcessor.isClosed) {
+          return;
+        }
+        _onError(GrpcError.unavailable('HTTPConnection connection-error'),
```
Comment (mosuem) [inline lib/src/client/transport/http_transport.dart:71]: See above, also `Http`

**Review by mosuem — CHANGES_REQUESTED** [review]
Thanks for the contribution! Tests are missing and a few smaller comments.

**Comment by mosuem** [comment]
Closing as this has been waiting for changes for a long time.

**Pull request closed without merging**